### PR TITLE
Update outputs of the CorrelatedJoin after the prunning 

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Closes https://github.com/crate/crate/issues/15901

Make sure that CorrelatedJoin keeps its outputs in sync with underlying tree's outputs after the prunning.

Needed to avoid the situation when a column is referenced but not collected anymore.

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
Fix is there, trying to find out a simpler reproduction and a simple example for a change entry. 

Cannot write to the change entry something that I added as a code comment in CorrelatedJoin, it's too vaque for users.

File: server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java [hunk 1]
```diff
@@ -162,12 +163,24 @@ public Void visitOuterColumn(OuterColumn outerColumn, Void ignored) {
         if (inputPlan == newInputPlan) {
             return this;
         }
+        // Original outputs can contain unused columns.
+        // We should update outputs after the prunning to avoid the edge case
+        // when a column is still referenced in CorrelatedJoin but not collected anymore after the prunning.
+        // See https://github.com/crate/crate/issues/1590 for details.
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java:169]: This comment is more internal and doesn't fit to a change entry.

Still don't have a simple example that I could use for a test/change entry. 

I confirmed that query in the reported issue works as expected with this change.

@mkleen, @mfussenegger may I get an early feedback on this? If direction is okay, I will continue looking for a simpler repro for a test/change entry.

Reverting https://github.com/crate/crate/pull/13736/ or https://github.com/crate/crate/pull/13895 is not an options as we don't want to get a performance regression. In general, it works because CorrelatedJoin only references actually collected columns.

Some workarounds mentioned in the issue force `unused_field` from the example query be collected.

File: libs/dex/src/main/java/io/crate/data/BiArrayRow.java [hunk 2]
```diff
@@ -29,22 +29,12 @@ public class BiArrayRow extends Row {
...
 
-    public BiArrayRow() {
-    }
 
     public BiArrayRow(Object[] firstCells, Object[] secondCells) {
         this.firstCells = firstCells;
         this.secondCells = secondCells;
     }
 
-    public void firstCells(Object[] cells) {
```
Comment (mkleen) [inline libs/dex/src/main/java/io/crate/data/BiArrayRow.java:40]: Is this related ?

Comment (BaurzhanSakhariev) [reply libs/dex/src/main/java/io/crate/data/BiArrayRow.java:40]: no, just a cleanup. Can revert it or do it in a separate commit

File: server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java (hunk 1 above)
Comment (mkleen) [reply server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java:169]: I think column-pruning should also be applied to the outputs of the CorrelatedJoin operator. This can be tested in isolation but testing `pruneOutputsExcept`.

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java:169]: Do you mean smth in addition to the current change (which is basically an implicit prunning) or replacing it with smth like

```
return new CorrelatedJoin(
            newInputPlan,
            selectSymbol,
            subQueryPlan,
            outputs
        ).pruneOutputsExcept(outputsToKeep); 
``` ?

Comment (mkleen) [reply server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java:169]: The change as it is applies to pruning to the outputs.

Comment (mkleen) [reply server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java:169]: So the current approach looks ok for me.

Comment (mfussenegger) [reply server/src/main/java/io/crate/planner/operators/CorrelatedJoin.java:169]: The direction is okay but I think the comments can be removed. That's just re-explaining what a proper `pruneOutputsExcept` implementation needs to do, and is documented on the itnerface.

File: libs/dex/src/main/java/io/crate/data/BiArrayRow.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply libs/dex/src/main/java/io/crate/data/BiArrayRow.java]: Pushed a dedicated commit

File: server/src/test/java/io/crate/planner/CorrelatedJoinPlannerTest.java [hunk 3]
```diff
@@ -82,4 +96,44 @@ public void test_correlated_subquery_cannot_be_used_in_having() {
         assertThatThrownBy(() -> e.plan(stmt))
             .hasMessage("Cannot use correlated subquery in HAVING clause");
     }
+
+
+    // Tracks https://github.com/crate/crate/issues/15901
+    @Test
+    public void test_correlated_join_outputs_reflects_input_plan_pruning() throws Exception {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/CorrelatedJoinPlannerTest.java:103]:
```suggestion
    public void test_prune_removes_unused_outputs_from_correlated_join() throws Exception {
```

File: server/src/test/java/io/crate/planner/CorrelatedJoinPlannerTest.java [hunk 4]
```diff
@@ -82,4 +96,44 @@ public void test_correlated_subquery_cannot_be_used_in_having() {
...
+        SQLExecutor e = SQLExecutor.of(clusterService)
+            .addTable("create table tbl (a int, b int)");
+        DocTableInfo tbl = e.resolveTableInfo("tbl");
+        var tableRelation = new DocTableRelation(tbl);
+
+        Symbol a = e.asSymbol("a");
+        Symbol b = e.asSymbol("b");
+
+        // A dummy relation containing at least one OuterColumn.
+        // Only columns, referenced by the outer columns, must be kept in CorrelatedJoin outputs.
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/CorrelatedJoinPlannerTest.java:113]:
```suggestion
```

File: server/src/test/java/io/crate/planner/CorrelatedJoinPlannerTest.java [hunk 5]
```diff
@@ -82,4 +96,44 @@ public void test_correlated_subquery_cannot_be_used_in_having() {
...
+        SelectSymbol selectSymbol = new SelectSymbol(
+            relation,
+            new ArrayType<>(DataTypes.INTEGER),
+            SelectSymbol.ResultType.SINGLE_COLUMN_SINGLE_VALUE,
+            true
+        );
+        CorrelatedJoin join = new CorrelatedJoin(collect, selectSymbol, null);
+        assertThat(join.outputs()).containsExactly(a, b, selectSymbol);
+        LogicalPlan pruned = join.pruneOutputsExcept(List.of(selectSymbol));
+        assertThat(pruned.outputs()).containsExactly(b, selectSymbol); // Column "a" is gone after the pruning.
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/CorrelatedJoinPlannerTest.java:137]:
```suggestion
        assertThat(pruned.outputs())
            .as("removes unused a")
            .containsExactly(b, selectSymbol);
```

**Review by mfussenegger — APPROVED** [review]
